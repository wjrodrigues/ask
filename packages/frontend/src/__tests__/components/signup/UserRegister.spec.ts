import { describe, it, expect } from "vitest";
import { shallowMount, mount } from "@vue/test-utils";

import UserRegisterVue from "@/components/signup/UserRegister.vue";
import { base_urls, basic_mount, faker, nock } from "@/__tests__/help";

describe("when render UserRegisterVue", () => {
  it("render form with all components", () => {
    const wrapper = shallowMount(
      UserRegisterVue,
      basic_mount({ props: {}, plugins: [] })
    );

    expect(wrapper.element).toMatchSnapshot();
  });

  it("validate required input", async () => {
    const wrapper = mount(
      UserRegisterVue,
      basic_mount({ props: {}, plugins: [] })
    );

    await wrapper.findAll("input").forEach(async (e) => e.setValue(""));
    await wrapper.trigger("change");

    const error_messages = await wrapper.findAll(".v-messages > div");

    expect(error_messages.length).toEqual(3);
    error_messages.map((msg) =>
      expect(msg.text()).toEqual("Campo é obrigatório")
    );
  });

  it("submit form button", async () => {
    const wrapper = mount(
      UserRegisterVue,
      basic_mount({ props: {}, plugins: [] })
    );

    expect(wrapper.find("button").attributes("disabled")).toBeUndefined;

    await wrapper.findAll("input").forEach(async (e) => e.setValue("Ask"));
    await wrapper.trigger("change");

    const data: any = wrapper.vm.$data;

    expect(data.form).toBeTruthy;
    expect(wrapper.find("button:disabled")).toBeNull;
    expect(wrapper.find("button").attributes("disabled")).toBeUndefined;
  });

  it("validates form submission errors", async () => {
    const wrapper = mount(
      UserRegisterVue,
      basic_mount({ props: {}, plugins: [] })
    );
    wrapper.find("[name='first_name']").setValue(faker.name.firstName());
    wrapper.find("[name='email']").setValue(faker.internet.email());
    wrapper.find("[name='password']").setValue(faker.internet.password());

    await wrapper.trigger("change");

    expect(wrapper.vm.errors).toEqual({
      first_name: "",
      email: "",
      password: "",
    });
    expect(wrapper.vm.form).toEqual(true);

    nock(base_urls.API_REGISTRATION)
      .post("/users")
      .reply(422, [{ email: ["já está em uso"] }]);

    await wrapper.find("form").trigger("submit");
  });
});
