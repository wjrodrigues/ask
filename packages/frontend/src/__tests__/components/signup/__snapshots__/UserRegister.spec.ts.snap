// Vitest Snapshot v1

exports[`when render UserRegisterVue > render form with all components 1`] = `
<v-card
  class="mx-auto"
  title="Cadastro do usuÃ¡rio"
  width="344"
>
  <v-container>
    <v-form
      modelvalue="false"
    >
      <v-text-field
        clearable=""
        label="Nome"
        rules="function () { [native code] }"
        type="text"
        variant="underlined"
      />
      <v-text-field
        clearable=""
        label="Sobrenome"
        type="text"
        variant="underlined"
      />
      <v-text-field
        clearable=""
        label="Email"
        rules="function () { [native code] }"
        type="email"
        variant="underlined"
      />
      <v-text-field
        clearable=""
        label="Senha"
        rules="function () { [native code] }"
        type="password"
        variant="underlined"
      />
      <v-card-actions>
        <v-btn
          block=""
          disabled="true"
          loading="false"
          rounded="pill"
          type="submit"
          variant="tonal"
        >
          Salvar
        </v-btn>
      </v-card-actions>
    </v-form>
  </v-container>
</v-card>
`;
